### PR TITLE
Issue #114

### PR DESCRIPTION
**Pull request by xUnholy**
Fixes:
- remove init functions
- include version in cli

### CODE REVIEW

File: internal/get/service.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+type getRunner struct {
+	cfg              *config.Config
+	ctx              context.Context
+	k8sClientset     *kubernetes.Clientset
+	k8sServiceClient clientV1.ServiceInterface
+}
+
+func Service(ctx context.Context, cfg *config.Config, service string) (*v1.Service, error) {
```
Comment (jbrekelmans) [inline internal/get/service.go]: `ctx` parameter is not used

File: cmd/down.go [hunk 2]
```diff
@@ -7,16 +7,23 @@ import (
 	"github.com/spf13/cobra"
 )
 
-var downCmd = &cobra.Command{
-	Use: "down",
-	Short: "Deletes the pods of the specified docker compose services. " +
-		"If all docker compose services would be deleted then the Kubernetes services are also deleted.",
-	Long: "destroy all pods and services",
-	Run:  downCommand,
+func SetupDownCli() *cobra.Command {
```
Comment (jbrekelmans) [inline cmd/down.go]: Doesn't need to be upper case I think

File: cmd/down.go [hunk 3]
```diff
@@ -7,16 +7,23 @@ import (
...
+			"If all docker compose services would be deleted then the Kubernetes services are also deleted.",
+		Long: "destroy all pods and services",
+		Run:  downCommand,
+	}
+	return downCmd
 }
 
 func downCommand(cmd *cobra.Command, args []string) {
-	cfg, err := upOrDownCommandCommon(cmd, args)
+	file, err := getFileFlag(cmd)
```
Comment (jbrekelmans) [inline cmd/down.go]: Should putShould move file back to `getCommandConfig`, so there's less duplication

File: cmd/get.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		Long:  "Print a detailed description of the selected resources, including related resources such as hostname or host IP.",
+		Run:   getCommand,
+	}
+	getCmd.PersistentFlags().StringP("format", "", "", "Return specified format")
+	return getCmd
+}
+
+func getCommand(cmd *cobra.Command, args []string) {
+	if len(args) == 0 {
+		log.Fatal("No Args Provided")
```
Comment (jbrekelmans) [inline cmd/get.go]: Can you add TODO reference to listing everything if no args are specified

File: internal/pkg/util/util.go [hunk 5]
```diff
@@ -113,3 +113,29 @@ func unescapeByte(input string, i int) (byte, error) {
...
+		for column, value := range row {
+			for len(maxValueWidthPerColumn) <= column {
+				maxValueWidthPerColumn = append(maxValueWidthPerColumn, 0)
+			}
+			if len(value) > maxValueWidthPerColumn[column] {
+				maxValueWidthPerColumn[column] = len(value)
+			}
+		}
+	}
+	sb := &strings.Builder{}
```
Comment (jbrekelmans) [inline internal/pkg/util/util.go]: Actually we can remove the `&` here and add `&` where we pass `sb` to `fmt.Fprintf`, this way we can remove a heap allocation if go isn't smart enough to figure this out

File: test/docker-compose.yml [hunk 6]
```diff
@@ -50,6 +50,7 @@ services:
       timeout: 5s
       retries: 5
     user: root
-# x-kube-compose:
-#   push_images:
-#     docker_registry: 'localhost:5000'
+
+x-kobercompose:
```
Comment (jbrekelmans) [inline test/docker-compose.yml]: What is `x-kobercompose`?

File: pkg/docker/compose/config/config.go [hunk 7]
```diff
@@ -295,7 +295,8 @@ func (c *configLoader) resolveExtends(
 // If files is an empty slice then the standard docker compose file locations (relative to the current working directory are considered).
 func New(files []string) (*CanonicalDockerComposeConfig, error) {
 	c := &configLoader{
-		environmentGetter: os.LookupEnv,
+		environmentGetter:     os.LookupEnv,
+		loadResolvedFileCache: map[string]*loadResolvedFileCacheItem{},
```
Comment (jbrekelmans) [inline pkg/docker/compose/config/config.go]: Nice bugfix

File: internal/pkg/get/get.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		return err
+	}
+	g.k8sClientset = k8sClientset
+	g.k8sServiceClient = g.k8sClientset.CoreV1().Services(g.cfg.Namespace)
+	return nil
+}
+
+func (g *getRunner) getK8sServiceResource(service *config.Service) (*v1.Service, error) {
+	options := &metav1.GetOptions{}
+	result, err := g.k8sServiceClient.Get(service.NameEscaped+"-"+g.cfg.EnvironmentID, *options)
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: `service.NameEscaped + cfg.EnvironmentID` is a common expression, we should add a function that computes this and call it where appropriate

File: internal/pkg/get/get.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+type KubeComposeService struct {
+	Service   string
+	Hostname  string
+	Namespace string
+	ClusterIP string
+}
+
+func Service(cfg *config.Config, serviceName string) (KubeComposeService, error) {
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: This name is not good, `GetServiceDetails` is better

File: internal/pkg/get/get.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	clientV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type getRunner struct {
+	cfg              *config.Config
+	k8sClientset     *kubernetes.Clientset
+	k8sServiceClient clientV1.ServiceInterface
+}
+
+type KubeComposeService struct {
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: Rename to `ServiceDetails`, we have about 50 thousand structs that include "service" in their name.

File: test/docker-compose.yml (hunk 6 above)
Comment (xUnholy) [reply test/docker-compose.yml]: Fixed

File: internal/pkg/get/get.go (hunk 9 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: Refactored to ServiceDetails, as the package is already get so it ends up being get.ServiceDetails

File: internal/pkg/get/get.go (hunk 8 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: Added getKubeServiceName() which takes service & cfg and returns the name

File: internal/pkg/get/get.go (hunk 9 above)
Comment (jbrekelmans) [reply internal/pkg/get/get.go]: I don't like this

File: internal/pkg/get/get.go (hunk 8 above)
Comment (jbrekelmans) [reply internal/pkg/get/get.go]: Should be `getK8sName()` to reflect that it can also be the name of a pod

File: internal/pkg/get/get.go (hunk 9 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: `func name will be used as get.GetServiceDetails by other packages, and that stutters; consider calling this ServiceDetails (golint)`

Comment (xUnholy) [reply internal/pkg/get/get.go]: Linter also does not like this naming convention.

File: cmd/down.go (hunk 3 above)
Comment (xUnholy) [reply cmd/down.go]: Done

File: internal/pkg/get/get.go (hunk 8 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: Done

File: internal/pkg/get/get.go (hunk 9 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: Done

File: internal/pkg/get/get.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	clientV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type getRunner struct {
+	cfg              *config.Config
+	k8sClientset     *kubernetes.Clientset
+	k8sServiceClient clientV1.ServiceInterface
+}
+
+type KubeComposeServiceDetails struct {
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: Just `ServiceDetails` without the `KubeCompose`

File: internal/pkg/get/get.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	k8sClientset, err := kubernetes.NewForConfig(g.cfg.KubeConfig)
+	if err != nil {
+		return err
+	}
+	g.k8sClientset = k8sClientset
+	g.k8sServiceClient = g.k8sClientset.CoreV1().Services(g.cfg.Namespace)
+	return nil
+}
+
+func (g *getRunner) getK8sServiceResource(service *config.Service) (*v1.Service, error) {
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: Move contents of `getK8sServiceResource` into `run`, then remove this function.

File: internal/pkg/get/get.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	return result, nil
+}
+
+func (g *getRunner) run(service *config.Service) (*v1.Service, error) {
+	err := g.initKubernetesClientset()
+	if err != nil {
+		return nil, err
+	}
+	result, err := g.getK8sServiceResource(service)
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: If statement is not needed based on this logic.

File: internal/pkg/get/get.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+type KubeComposeServiceDetails struct {
+	Service   string
+	Hostname  string
+	Namespace string
+	ClusterIP string
+}
+
+func GetServiceDetails(cfg *config.Config, serviceName string) (KubeComposeServiceDetails, error) {
+	composeService := KubeComposeServiceDetails{}
```
Comment (jbrekelmans) [inline internal/pkg/get/get.go]: GetServiceDetails should be small, similar to `up.Run` and `down.Run`, like this:
```
return (&getRunner{cfg: cfg, service: service}).run()
```

File: cmd/down.go [hunk 15]
```diff
@@ -7,16 +7,19 @@ import (
 	"github.com/spf13/cobra"
 )
 
-var downCmd = &cobra.Command{
-	Use: "down",
-	Short: "Deletes the pods of the specified docker compose services. " +
-		"If all docker compose services would be deleted then the Kubernetes services are also deleted.",
-	Long: "destroy all pods and services",
-	Run:  downCommand,
+func setupDownCli() *cobra.Command {
```
Comment (jbrekelmans) [inline cmd/down.go]: Rename to `newXXXCommand`, e.g. here: `newDownCommand`

Comment (xUnholy) [reply cmd/down.go]: Done

File: internal/pkg/get/get.go (hunk 12 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: done

File: internal/pkg/get/get.go (hunk 13 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: done

File: internal/pkg/get/get.go (hunk 11 above)
Comment (xUnholy) [reply internal/pkg/get/get.go]: Done

**Pull request merged**